### PR TITLE
[#10243][feat] Add TRT-LLM attention backend to AutoDeploy

### LINKED ISSUE #10243: [Feature]:  Add TRTLLM as an attention backend for Auto Deploy
### 🚀 The feature, motivation and pitch

TRTLLM attention backend gives better perf for some models. Until flashInfer catches up, it'll be good to use TRTLLM backend to have on-par perf. So we need to add this backend option.

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by MrGeva**
fixes #10243 

Adds a new trtllm attention backend for Auto-Deploy that wraps TRT-LLM's optimized thop.attention kernel.

  • New backend implementation: tensorrt_llm/_torch/auto_deploy/custom_ops/trtllm_attention.py
    • Registers TrtllmAttention in AttentionRegistry under "trtllm".
    • Implements a cached-attention custom op (trtllm_attention_mha_with_cache) that:
      • Accepts Q/K/V in AD’s bsnd layout.
      • Builds fused QKV ([num_tokens, (n_heads + 2*n_kv_heads)*head_dim]) as required by thop.attention.
      • Consumes AD’s paged KV cache in HND layout ([num_blocks, 2, n_kv_heads, tokens_per_block, head_dim]).
      • Translates AD metadata (batch_info, cu_seqlen, page tables) into TRT-LLM FMHA metadata tensors (sequence_length, context_lengths, request_types, kv_cache_block_offsets, etc.).

  KV cache / pool integration

  • Uses KVPagedResourceHandler-compatible cache initializer (TrtllmKVResourceHandler) so the standard KVCacheManager is used.
  • Plumbs KV cache pool pointers and layer→pool mapping from the cache manager into the TRT-LLM backend (needed by thop.attention for paged cache addressing).

  CUDA graph correctness + performance

  • Adds a host_prepare path that runs outside the captured graph before each forward/replay:
    • Writes only “active” slices of metadata and uses pinned host buffers to avoid per-iteration allocations.
  • Introduces shared batch-info state (CudaGraphState.set/get_batch_info) so hybrid-model custom ops (Mamba/SSM + causal conv) can obtain the current batch composition during CUDA-graph replay (where batch_info_host.tolist() can be
    stale).

  Hybrid model support (Nano/Nemotron)

  • Updates Mamba-related custom ops to use the shared batch-info when available, ensuring correct state gather/scatter and preventing replay-time stale-dimension bugs for hybrid models.

  Quantization handling

  • Supports FP8 KV cache by setting QuantMode.FP8_KV_CACHE for thop.attention.
  • Adds explicit errors for unsupported KV-cache formats (e.g., INT8 / NVFP4 packed KV cache) so failures are clear instead of silent misbehavior.

  Tests + CI wiring

  • Parametrizes AutoDeploy accuracy tests to include attn_backend="trtllm".
  • Updates CI test lists to run trtllm-only for AutoDeploy coverage where requested, and updates waives accordingly.

## Release Notes

* **New Features**
  * Added TRT-LLM as a new attention backend option, expanding backend choices for model inference

* **Improvements**
  * Enhanced CUDA graph capture stability through refined warm-up sequences
  * Improved batch dimension handling during CUDA graph replay for Mamba operations

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Comment by MrGeva** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1"

**Comment by MrGeva** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1"

File: tensorrt_llm/_torch/auto_deploy/custom_ops/trtllm_attention.py [hunk 1]
```diff
@@ -0,0 +1,1387 @@
...
+    cache_loc: torch.Tensor,
+    last_page_len: torch.Tensor,
+    last_page_len_host: torch.Tensor,
+    seq_len_with_cache_host: torch.Tensor,
+    state: TrtllmLayerState,
+    kv_cache: torch.Tensor,
+    ad_pool_pointers: Optional[torch.Tensor] = None,
+    ad_pool_mapping: Optional[torch.Tensor] = None,
+) -> Tuple[torch.Tensor, ...]:
+    """Prepare TRT-LLM metadata from AD metadata.
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/trtllm_attention.py]: this will have to go. Two places where it should go:

1. Use existing fields in attention interface that are available
2. Add fields to attention interface if it makes sense
3. Use one or both of the existing standardized prepare_metadata interfaces (host or device side)

File: tensorrt_llm/_torch/auto_deploy/custom_ops/trtllm_attention.py [hunk 2]
```diff
@@ -0,0 +1,1387 @@
...
+            return
+
+        self._gpu_cu_pages = torch.zeros(max_seqs + 1, dtype=torch.long, device="cuda")
+        self._gpu_page_positions = torch.arange(max_pages, dtype=torch.long, device="cuda")
+        self._gpu_seq_idx = torch.zeros(max_pages, dtype=torch.long, device="cuda")
+        self._gpu_page_idx = torch.zeros(max_pages, dtype=torch.long, device="cuda")
+        self._gpu_base_offset = torch.zeros(max_pages, dtype=torch.int32, device="cuda")
+        self._gpu_buffers_initialized = True
+
+    def _update_pool_pointers(
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/trtllm_attention.py]: I don't understand the need for these separate pool pointers

**Review by lucaslie — COMMENTED** [review]
I think we need to take a deep look at the attention metadata and dissect one-by-one to remove and reduce all the extra metadata wrappers in the trtllm_attention integration

**Comment by MrGeva** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1" --disable-fail-fast

**Review by lucaslie — CHANGES_REQUESTED** [review]

File: tensorrt_llm/_torch/auto_deploy/custom_ops/trtllm_attention.py [hunk 3]
```diff
@@ -0,0 +1,1387 @@
...
+        return {
+            "kv_cache": TrtllmKVResourceHandler(
+                num_kv_heads=num_kv_heads,
+                head_dim=head_dim,
+                dtype=dtype,
+                layer_idx=layer_idx,
+                trtllm_config=_trtllm_config,
+                cache_config=cache_config,
+            ),
+            "workspace_buffer": TrtllmWorkspaceResourceHandler(),
```
[the remark this answers is not in the document]
Comment (MrGeva) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/trtllm_attention.py]: its allocating a single workspace for all layers already, but I will avoid using the ResourceHandler mechanism for the allocation

**Comment by MrGeva** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1"

**Comment by MrGeva** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1"

**Comment by MrGeva** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1"

**Review by galagam — COMMENTED** [review]
partial review

**Comment by MrGeva** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1"

**Comment by MrGeva** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1"

File: tensorrt_llm/_torch/auto_deploy/custom_ops/attention/trtllm_attention.py [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+    # Input is always [bs, 1] (generate-only) or [1, total_seq_len] (prefill/mixed),
+    # so b * s == num_tokens always holds.
+    q_shape_og = q.shape
+    q_flat = q.reshape(num_tokens, num_heads * head_dim)
+    k_flat = k.reshape(num_tokens, num_kv_heads * head_dim)
+    v_flat = v.reshape(num_tokens, num_kv_heads * head_dim)
+    qkv_fused = torch.cat([q_flat, k_flat, v_flat], dim=-1).contiguous()
+
+    # Prepare output
+    output = torch.empty(num_tokens, num_heads * head_dim, dtype=q.dtype, device=q.device)
```
[the remark this answers is not in the document]
Comment (MrGeva) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/attention/trtllm_attention.py]: It does not happen in cudagraph replay, so no ROI

**Review by lucaslie — APPROVED** [review]
👏

**Comment by MrGeva** [comment]
/bot skip --comment "trivial conflict only in the test list from last passing CI"
